src/Nethermind/Nethermind.Trie/PatriciaTree.cs

```diff
@@ -1133,18 +1157,19 @@ TrieNode secondExtension
             return ref traverseContext.UpdateValue;
         }
 
-        private ref readonly CappedArray<byte> TraverseNext(TrieNode next, scoped in TraverseContext traverseContext, int extensionLength)
+        private ref readonly CappedArray<byte> TraverseNext(TrieNode next, scoped in TraverseContext traverseContext, scoped ref TreePath path, int extensionLength)
         {
             // Move large struct creation out of flow so doesn't force additional stack space
             // in calling method even if not used
             TraverseContext newContext = traverseContext.WithNewIndex(traverseContext.CurrentIndex + extensionLength);
-            return ref TraverseNode(next, in newContext);
+            return ref TraverseNode(next, in newContext, ref path);
         }
 
         private readonly ref struct TraverseContext
         {
             public readonly ref readonly CappedArray<byte> UpdateValue;
             public readonly ReadOnlySpan<byte> UpdatePath;
+            public readonly ref readonly TreePath UpdatePathTreePath;
```

Comment:
1. Why keep the `UpdatePath` separate from `UpdatePathTreePath`?
2. In a few call sites, if not all of them, the additional ref is passed beside the traverse context to pass the path in the tree. Why not stick it to the traverse context so that traversal methods are 1 params shorter, like they used to be?

Reply:
So the `UpdatePathTreePath` was added later because `ConnectNodes` needed it and `UpdateValue` -> TreePath nibble conversion is actually a significant bottleneck. 

As for why the TreePath is not in TraverseContext in the first place, if I'm not mistaken, I tried it and something broke, and TraverseContext is cloned during traversal so the whole path is a bit heavy there, I thought just adding `ref TreePath` makes it more consistent everywhere and TreePath mutation's optimization turns out to have a measurable impact.

Reply:
Oh, I see. I'd add this reasoning in there as a comment that the context is cloned while the path is used as a builder that is passed around. Then, I think, we can mark it as solved.